### PR TITLE
Use spec.format to generate module vars

### PR DESCRIPTION
**Pull request by scheibelp**
This replaces modules.py current variable substitution with calls to spec.format. The goal is to use the same approach where possible so users do not need to learn multiple substitution formats. This includes:

* generation of module file names
* generation of environment variable names and values

This also includes the addition of 2 fields to spec.format: prefix and hash.

### CODE REVIEW

**Comment by scheibelp** [comment]
@alalazo @mamelara I'm curious if you have thoughts/objections on this. 

@tgamblin mentioned that referring spec details like {compiler.name} may be preferable to {compilername} as spec.format currently requires. That is not currently part of this PR

I assume this affects additional users - this would break current modules.yaml so anyone who is making use of modules would be affected - so if you know who they are it would be great if you could add them.

Thanks!
Peter

File: lib/spack/spack/modules.py [hunk 1]
```diff
@@ -192,8 +192,6 @@ def parse_config_options(module_generator):
     #####
 
     # Automatic loading loads
-    module_file_actions['hash_length'] = module_configuration.get(
-        'hash_length', 7)
```
Comment (alalazo) [inline lib/spack/spack/modules.py]: This, together with other changes in the PR, removes any effect of the `hash_length` keyword in `modules.yaml` and adds the possibility of specifying the hash from the naming scheme. I don't think this is a good idea 

I suppose many people are just fine with the default naming scheme and would like to adjust only the hash length. In the current state of things they can do it adding the line:
```yaml
hash_length: <some-length>
```
to the configuration file. After this PR they will need to write : 
```yaml
'naming_scheme': '${PACKAGE}-${VERSION}-${COMPILERNAME}-${HASH:<some-length>}',
```
which is more boiler-plate for the same amount of information.

File: lib/spack/spack/modules.py [hunk 2]
```diff
@@ -500,7 +478,7 @@ class Dotkit(EnvModule):
     autoload_format = 'dk_op {module_file}\n'
 
     default_naming_format = \
-        '{name}-{version}-{compiler.name}-{compiler.version}'
+        '${PACKAGE}-${VERSION}-${COMPILERNAME}-${COMPILERVER}-${HASH:7}'
```
Comment (alalazo) [inline lib/spack/spack/modules.py]: See above : I would not insert the hash in the naming format

File: lib/spack/spack/modules.py [hunk 3]
```diff
@@ -275,48 +277,23 @@ def naming_scheme(self):
...
-        name = naming_scheme.format(**naming_tokens)
+        name = self.spec.format(self.naming_scheme)
         # Not everybody is working on linux...
         parts = name.split('/')
         name = join_path(*parts)
         # Add optional suffixes based on constraints
         configuration, _ = parse_config_options(self)
         suffixes = [name]
+        #TODO: leave this for now - just replace the default naming
+        #scheme with a spec.format-compatible one
```
Comment (alalazo) [inline lib/spack/spack/modules.py]: Just in case : I believe hashes should always come last in the name, i.e. : 
```
<rest-of-the-name>-<suffixes>-<hash>
```
and not : 
```
<rest-of-the-name>-<hash>-<suffixes>
```

File: lib/spack/spack/modules.py [hunk 4]
```diff
@@ -543,7 +521,7 @@ class TclModule(EnvModule):
     prerequisite_format = 'prereq {module_file}\n'
 
     default_naming_format = \
-        '{name}-{version}-{compiler.name}-{compiler.version}'
+        '${PACKAGE}-${VERSION}-${COMPILERNAME}-${COMPILERVER}-${HASH:7}'
```
Comment (alalazo) [inline lib/spack/spack/modules.py]: Same here

File: lib/spack/spack/modules.py [hunk 5]
```diff
@@ -674,21 +651,22 @@ class LmodModule(EnvModule):
...
-    hierarchy_tokens = hierarchy_tokens + ['mpi', 'compiler']
-
     def __init__(self, spec=None):
         super(LmodModule, self).__init__(spec)
+
+        self.configuration = CONFIGURATION.get('lmod', {})
+        hierarchy_tokens = self.configuration.get('hierarchical_scheme', [])
+        # TODO : Check that the extra hierarchy tokens specified in the
+        # TODO : configuration file are actually virtual dependencies
+        self.hierarchy_tokens = hierarchy_tokens + ['mpi', 'compiler']
```
Comment (alalazo) [inline lib/spack/spack/modules.py]: Thanks for fixing this.: `hierarchy_tokens` was meant to be set during init + this will solve weird issues when mocking up values in tests.

File: lib/spack/spack/spec.py [hunk 6]
```diff
@@ -2129,10 +2129,12 @@ def format(self, format_string='$_$@$%@+$+$=', **kwargs):
             ${OPTIONS}       Options
             ${ARCHITECTURE}  Architecture
             ${SHA1}          Dependencies 8-char sha1 prefix
+            ${HASH:len}      DAG hash with optional length specifier
 
             ${SPACK_ROOT}    The spack root directory
             ${SPACK_INSTALL} The default spack install directory,
                              ${SPACK_PREFIX}/opt
+            ${PREFIX}        The package prefix
```
Comment (alalazo) [inline lib/spack/spack/spec.py]: I tried to search for uses of full-string substitutions in the whole code base. Am I wrong in saying that there's none ?

**Comment by alalazo** [comment]
I'll ping @glennpj and @robertdfrench as their comments were quite helpful to me for other module-file related PRs :smile:

**Comment by alalazo** [comment]
> @tgamblin mentioned that referring spec details like {compiler.name} may be preferable to {compilername} as spec.format currently requires.

I consider this a minor point, but I tend to agree with @tgamblin . In using the current `Spec.format` we are effectively substituting a python-like string formatting with a bash-like one.

**Comment by alalazo** [comment]
Another thing : I just saw that I forgot to port this a4f15011ed6fcceffc5fc9d8fc3f712bd7c27f6b from our production branch. Please let me know if you want this added in develop right now, in which case I can submit a PR.

**Comment by scheibelp** [comment]
>I agree with the first aim and not really with the second because the full-string syntax is never used in the code-base

When you say you are OK with 1 but not 2, are you suggesting that the module configuration settings are translated into spec.format strings? For example hash_length: 7 translates into appending {HASH:7} to the end? 

I think you're right about full-string syntax not being used so far, but I think that it is clearer from a user perspective. At some point I think we were interested in exposing the spec.format syntax to users: it is the basis of specifying name projections when creating RPMs from Spack packages (although that is not merged yet). Is there a reason you would not want to expose it to the user?  The full-string spec.format syntax looks similar to the syntax that is currently used in modules.py

> I tried to search for uses of full-string substitutions in the whole code base. Am I wrong in saying that there's none ?

I think you are right. Is there an issue if there are no other cases? I felt SPACK_ROOT and SPACK_PREFIX set a precedent for inserting paths into the format string.

> Please let me know if you want https://github.com/LLNL/spack/commit/a4f15011ed6fcceffc5fc9d8fc3f712bd7c27f6b added in develop right now, in which case I can submit a PR.

I can add that in here (or modify as needed) - thanks for linking it! That being said if you want it in develop sooner than this PR I don't mind merging that in.

**Comment by scheibelp** [comment]
The latest commit makes use of hash_len instead of {HASH} as part of spec.format. It should also support setting hash_len=0 for LmodModules as in https://github.com/LLNL/spack/commit/a4f15011ed6fcceffc5fc9d8fc3f712bd7c27f6b

(I should add a unit test for this before taking off [WIP])

**Comment by alalazo** [comment]
> I think you're right about full-string syntax not being used so far, but I think that it is clearer from a user perspective. At some point I think we were interested in exposing the spec.format syntax to users: it is the basis of specifying name projections when creating RPMs from Spack packages (although that is not merged yet); it is also anticipated that this format would be useful for specifying projections of Spack package prefixes to a simplified symlink tree (to make user navigation of installed packages via filesystem easier). Is there a reason you would not want to expose it to the user? The full-string spec.format syntax looks similar to the syntax that is currently used in modules.py

Ok, then I think I was just missing some context, as I am not following the RPM pull request close enough. Basically my reasoning was : 

1. full-string spec syntax and current syntax are of equivalent complexity (there's no benefit in using one instead of the other)
1. current syntax is in use in `modules.yaml`, while full-string spec syntax is not exposed to user anywhere

It just didn't make much sense to me to substitute *in a configuration file* a syntax that is already in use with an equivalent one that brings no benefit.

This point of course doesn't hold if other planned merges will expose full-string spec syntax to user : at that point we will have two different ways of saying the same thing (e.g. `${COMPILERNAME}` vs. `{compiler.name}`) and one must be pruned.

> The latest commit makes use of hash_len instead of {HASH} as part of spec.format. It should also support setting hash_len=0 for LmodModules as in a4f1501

Thanks! The only concern I have is about : 
```python
# Path part for a provider
if name in self.hierarchy_tokens:
    suffixes.append(value.dag_hash(length=7))
 
value.hash = '-'.join(suffixes)
if value.hash:
    return self.path_part_with_hash.format(token=value)
return self.path_part_without_hash.format(token=value)
```
which doesn't seem to be ported so far. This part unconditionally keeps the hashes in the directory names. The use case for this is providing two flavors of the same `mpi` or `lapack` and setting `hash_length: 0` : in that case you want that things that are compiled against, say, `openblas+openmp` do not show up in the same folder as `openblas~openmp` (which would be the case if you remove hashes and use the same `openblas` version)

If you agree, when you are ready with this PR, I can try it out on a couple of edge cases we encountered in production.  If need be I can submit a PR on your branch with additional test cases or logic.

**Comment by scheibelp** [comment]
>The only concern I have is about : 

I think I see that while hash should be taken care of I left out the suffixes for lmod modules. Today's commits add that in.

**Comment by scheibelp** [comment]
I have rebased to clean up. I have added unconditional hashes for directory names in LmodModules and also added a unit test for that.

**Comment by alalazo** [comment]
@tgamblin This will be part of v0.10, right ?

**Comment by tgamblin** [comment]
seems like it really should be.  0.10 is going to be a major release.  Maybe it should just be 1.0 :)

**Comment by tgamblin** [comment]
@glennpj @robertdfrench: are you ok with these changes?  This is a minor update for consistency that affects `modules.yaml` syntax.

**Comment by glennpj** [comment]
@tgamblin 
Yep, this looks fine to me.
